infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/multiuser/oauth/IdentityProviderConfigFactory.java

```diff
@@ -103,70 +90,123 @@ public IdentityProviderConfigFactory(
-          workspaceRuntimeProvider.get().getRuntimeContext(workspaceId);
-      workspaceOwnerId = context.map(c -> c.getIdentity().getOwnerId()).orElse(null);
+    if (oauthIdentityProvider == null) {
+      LOG.debug("OAuth Provider is not configured, default config is used.");
+      return defaultConfig;
     }
 
-    if (oauthIdentityProvider != null
-        && subject != Subject.ANONYMOUS
-        && (workspaceOwnerId == null || subject.getUserId().equals(workspaceOwnerId))) {
-      try {
-        KeycloakTokenResponse keycloakTokenInfos =
-            keycloakServiceClient.getIdentityProviderToken(oauthIdentityProvider);
-        if ("user:full".equals(keycloakTokenInfos.getScope())) {
-          return new OpenShiftConfigBuilder(OpenShiftConfig.wrap(defaultConfig))
-              .withOauthToken(keycloakTokenInfos.getAccessToken())
-              .build();
-        } else {
-          throw new InfrastructureException(
-              "Cannot retrieve user Openshift token: '"
-                  + oauthIdentityProvider
-                  + "' identity provider is not granted full rights: "
-                  + oauthIdentityProvider);
-        }
-      } catch (UnauthorizedException e) {
-        LOG.error("cannot retrieve User Openshift token from the identity provider", e);
-
-        throw new InfrastructureException(messageToLinkAccount);
-      } catch (BadRequestException e) {
-        LOG.error(
-            "cannot retrieve User Openshift token from the '"
-                + oauthIdentityProvider
-                + "' identity provider",
-            e);
-        if (e.getMessage().endsWith("Invalid token.")) {
-          throw new InfrastructureException(
-              "Your session has expired. \nPlease "
-                  + "<a href='javascript:location.reload();' target='_top'>"
-                  + "login"
-                  + "</a> to Che again to get access to your Openshift account");
-        }
-        throw new InfrastructureException(e.getMessage(), e);
-      } catch (Exception e) {
-        LOG.error(
-            "cannot retrieve User Openshift token from the  '"
+    if (subject == Subject.ANONYMOUS) {
+      LOG.debug(
+          "OAuth Provider is configured but default subject is anonymous, default config is used.");
+      return defaultConfig;
+    }
+
+    if (workspaceId == null) {
+      LOG.debug(
+          "OAuth Provider is configured and this request is not related to any workspace. OAuth token will be retrieved.");
+      return personalizeConfig(defaultConfig);
+    }
+
+    Optional<RuntimeContext> context =
+        workspaceRuntimeProvider.get().getRuntimeContext(workspaceId);
+    if (!context.isPresent()) {
+      // TODO Not sure about this tricky part.
+      // there is no cached info for this workspace in workspace API
+      // it really
+      LOG.debug(
+          "Runtime Content for the specified workspace '%s' is not found. OAuth token will be retrieved.");
```

Comment:
`Runtime Context`?

Reply:
Good catch! The message if rewritten now
`There is no runtime context for the specified workspace '%s'. It's the first workspace related call, so context is personalized with OAuth token.`